### PR TITLE
[RLlib] Implement `TorchPolicy.export_model`.

### PR DESCRIPTION
**Pull request by sven1977**
Implement `TorchPolicy.export_model`.

This method is currently missing and users get an NotImplementedError when trying to call it on any TorchPolicy.
This PR provides a solution storing the Policy's model as a TorchScript.
The only remaining problem is for non-RNNs: A fake state in tensor (any tensor is fine) must be provided when calling the TorchScript model. This is due to torch.jit not handling the otherwise empty internal states list that is returned by RLlib's ModelV2s.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/torch_policy.py [hunk 1]
```diff
@@ -598,9 +599,32 @@ def optimizer(
...
+        Creates a TorchScript model and saves it.
+
+        Args:
+            export_dir (str): Local writable directory or filename.
         """
-        raise NotImplementedError
+        dummy_inputs = self._lazy_tensor_dict(self._dummy_batch.data)
+        # Provide dummy state inputs if not an RNN (torch cannot jit with
+        # returned empty internal states list).
+        if "state_in_0" not in dummy_inputs:
```
Comment (michaelzhiluo) [inline rllib/policy/torch_policy.py:612]: Why does torch jit require state_in even if the model is not an RNN?

**Review by michaelzhiluo — CHANGES_REQUESTED** [review]

File: rllib/policy/torch_policy.py [hunk 2]
```diff
@@ -598,9 +599,32 @@ def optimizer(
...
+        while "state_in_{}".format(i) in dummy_inputs:
+            state_ins.append(dummy_inputs["state_in_{}".format(i)])
+            i += 1
+        seq_lens = dummy_inputs["seq_lens"]
+        dummy_inputs = {k: dummy_inputs[k] for k in dummy_inputs.keys()}
+        traced = torch.jit.trace(self.model,
+                                 (dummy_inputs, state_ins, seq_lens))
+        if not os.path.exists(export_dir):
+            os.makedirs(export_dir)
+        file_name = os.path.join(export_dir, "model.pt")
```
Comment (michaelzhiluo) [inline rllib/policy/torch_policy.py:626]: Also enable user to customize model name^

**Review by michaelzhiluo — CHANGES_REQUESTED** [review]

File: rllib/policy/torch_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py:612]: Agree, this is a total hack. Torch requires the output to be some tensor (or nested struct of tensors), but NOT an empty list :/ That's why we need to fake it here. One more reason to keep **thinking** about a possible new ModelV3 API :)

File: rllib/policy/torch_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py:626]: Yeah, was thinking about this, too. The problem is: I didn't want to change the method's signature, which is:
`export_model(self, export_dir)`. For TF, this is ok b/c a TF model export will produce many files (inside the export_dir).
For Torch, it's just a single file.
But yes, we could add an optional arg (filename=None) to the torch method (and then make the base's signature: `export_dir, **kwargs`).

File: rllib/policy/torch_policy.py (hunk 1 above)
Comment (eadlam) [reply rllib/policy/torch_policy.py:612]: torch jit doesn't require state_in, it's the self.model that requires it.

**Review by michaelzhiluo — APPROVED** [review]

**Comment by michaelzhiluo** [comment]
Once conflict is done, we can merge it ^^

**Pull request merged**